aehmc/proposals.py

```diff
@@ -0,0 +1,95 @@
+from typing import Callable, Tuple
+
+import aesara.tensor as aet
+import numpy as np
+from aesara.tensor.random.utils import RandomStream
+from aesara.tensor.var import TensorVariable
+
+from aehmc.integrators import IntegratorStateType
+
+ProposalStateType = Tuple[
+    IntegratorStateType, TensorVariable, TensorVariable, TensorVariable
+]
```

Comment:
`ProposalStateType` and `IntegratorStateType` should be `namedtuple`s or `dataclass`es.  That change&mdash;along with some corresponding changes from indexing to field names&mdash;will make this code a lot easier to follow.

Reply:
I agree but we're going to change something on the `aesara` side to be able to use them in `scan`.

Reply:
I'm essentially waiting for us to replace the C backend with Numba as the default.  After that, we can very easily add new `Type`s for `tuple`s, `list`s, `str`s, `namedtuple`s, etc.  We could do that now, but those types would need to be in Python&mdash;and, thus, not as efficient to use&mdash;or C, and that would involve some very annoying and often involved CPython-via-string-templating work.

In the meantime, we can write our code to use `namedtuples` and then make the change when those types are available.